### ISSUE TITLE
Python Reference Count Problems

### ISSUE THREAD

**Issue by inflector** [opencog/opencog#1303]
In both PythonEval.cc and PythonModule.cc

I fixed the ones in `PythonEval::init`, `PythonEval::apply`, `PythonEval::apply_script`, but there are still a lot of calls like:

`PyDict_SetItem(PyModule_GetDict(mod), PyBytes_FromString("ATOMSPACE"), this->getPyAtomspace());`

in PythonEval.cc

The above call to `PyDict_SetItem` will cause the dictionary Python objects to never get cleaned up because the reference counts are never decremented since the `PyObject*` is passed on the stack and  `Py_DECREF` isn't called on the object returned by the Python C Library calls which return `PyObject*` objects.

This is how the above should be done in order not to leak Python objects:

```
    PyObject *pyDict, *pyAtomSpaceKey, *pyAtomSpace;
    pyDict = PyModule_GetDict(this->pyRootModule);
    pyAtomSpaceKey = PyBytes_FromString("ATOMSPACE");
    pyAtomSpace = this->getPyAtomspace();
    PyDict_SetItem(pyDict, pyAtomSpaceKey, pyAtomSpace);
    Py_DECREF(pyDict);
    Py_DECREF(pyAtomSpaceKey);
    Py_DECREF(pyAtomSpace);
```

**Comment by inflector** [comment]
Need to go through the all the functions that return PyObject* one by one to double-check the borrowed versus new reference status. Still finding problems.

**Comment by inflector** [comment]
Stolen references are another issue that could lead to crashes. Decrement a reference that you naively assume you own but that some subsequent Python API call "stole" and you will cause a crash. Might be on the decrement in your code, if you are lucky, or perhaps some indeterminate time later when Python decrements it down to zero. So you have to check every single API call where you pass in PyObject* pointers to verify whether or not that call "steals" references. Do they all steal them? No, that would be too easy. Just a few. Yikes! 

The Python C api is the most confusing and fragile WTF API I have ever used. No sanity checks on the inputs, at all. Pass a NULL where one isn't expected? It segfaults. They just expose their internal API for public use.

Inconsistent reference counting rules that only lead to confusion saves what? Less than 8 instructions on memory that's going to be in the L1 cache. Seriously, look at how few instructions must be in the Py_INCREF macro when you're not running a debugging version of Python:
```
((PyObject*)(op))->ob_refcnt++)
```
So you've got a pointer deference with offset with an increment. What's that 2 to 3 instructions (depending on the processor) on memory that's going to be in the cache anyway? 

Py_DECREF is just:
```
if (--((PyObject*)(op))->ob_refcnt==0)
        _Py_Dealloc((PyObject *)(op));
```
So you've got a pointer deference with offset, a decrement, and a compare. Again, 3 or 4 instructions. If they added a Py_INCREF to eliminate "stolen" references, then the deallocate is not going to happen anyway so you're not saving that.

That's all they're saving by the inconsistency. They could make the calls completely consistent by simply doing away with borrowed references by adding a Py_INCREF and stolen references with the same Py_INCREF.  In each case, that would result in the caller doing a corresponding Py_DECREF they don't now need to do. 

So they're saving us 5 to 7 instructions per call on a few of the functions at the cost of pervasive confusion, countless articles, and rants like this one. :-) That might make fine sense for the internals of Python, but hell this is for running Python code after all, if we were that worried about a few instruction we'd be writing in C or C++ and gain an order of magnitude or two.

They should wrap all this internal and inconsistent crap into a set of public API functions that always return references that are owned and need to be decremented. And they should never steal them. Then the memory management would be simple, any PyObject*s returned by the API need to be decremented when you are done with them, that's it. No tricks.

**Comment by linas** [comment]
In that case, we're good, as long as we can keep you from looking under the overs of what goes on in the atomspace.

**Comment by inflector** [comment]
Well AtomSpace should be messy, it's fairly new and still in flux.  :-)

The Python C API is used by thousands of developers all across the world. There are tens of articles on the internet talking confusedly about how to use it and how hard it is. And it's been in use for what? 15 to 20 years?

I used an open-source Visual Basic-esque language called Bee Basic that was written by one guy that had much better and less confusing interfaces more than 10 years ago. And I wrote developer tools for 10 years in the 80s and 90s before the days when open-source took off, so I'm kind of particular about these sorts of things.
